packages/core/src/render3/instructions/styling.ts

```diff
@@ -304,323 +136,221 @@ export function ɵɵstyleMap(styles: {[styleName: string]: any} | NO_CHANGE | nu
-    tNode.flags |= TNodeFlags.hasInitialStyling;
-  }
-
-  return hasAdditionalInitialStyling;
-}
-
-function updateRawValueOnContext(
-    context: TStylingContext | StylingMapArray | string, value: string) {
-  const stylingMapArr = getStylingMapArray(context) !;
-  stylingMapArr[StylingMapArrayIndex.RawValuePosition] = value;
-}
-
-function getStylesContext(tNode: TNode): TStylingContext {
-  return getContext(tNode, false);
-}
-
-function getClassesContext(tNode: TNode): TStylingContext {
-  return getContext(tNode, true);
+export function hasStylingInputShadow(tNode: TNode, isClassBased: boolean) {
+  return (tNode.flags & (isClassBased ? TNodeFlags.hasClassInput : TNodeFlags.hasStyleInput)) !== 0;
 }
 
 /**
- * Returns/instantiates a styling context from/to a `tNode` instance.
- */
-function getContext(tNode: TNode, isClassBased: boolean): TStylingContext {
-  let context = isClassBased ? tNode.classes : tNode.styles;
-  if (!isStylingContext(context)) {
-    const hasDirectives = isDirectiveHost(tNode);
-    context = allocTStylingContext(context as StylingMapArray | null, hasDirectives);
-    if (ngDevMode) {
-      attachStylingDebugObject(context as TStylingContext, tNode, isClassBased);
-    }
-
-    if (isClassBased) {
-      tNode.classes = context;
-    } else {
-      tNode.styles = context;
+* Flushes styling into DOM element from the bindings.
+*
+* The function starts at `LFrame.stylingBindingChanged` and computes new styling information from
+* the bindings progressing towards the tail of the list. At the end the resulting style is written
+* into the DOM Element.
+*
+* This function is invoked from:
+* 1. Template `advance` instruction.
+* 2. HostBinding instruction.
+*/
+function flushStylingOnElementExit() {
+  ngDevMode && assertEqual(
+                   getStyleBindingChanged() > 0 || getClassBindingChanged() > 0, true,
+                   'Only expected to be here if binding has changed.');
+  ngDevMode &&
+      assertEqual(
+          getCheckNoChangesMode(), false, 'Should never get here during check no changes mode');
+  const lView = getLView();
+  const tView = lView[TVIEW];
+  const tData = tView.data;
+  const elementIndex = getSelectedIndex() + HEADER_OFFSET;
+  const tNode = tData[elementIndex] as TNode;
+  const renderer = lView[RENDERER];
+  const element = unwrapRNode(lView[elementIndex]) as RElement;
+
+  const classBindingIndex = getClassBindingChanged();
+  if (classBindingIndex > 0) {
+    const classLastWrittenValueIndex = getTStylingRangeTail(tNode.classBindings) + 1;
+    ngDevMode &&
+        assertGreaterThan(
+            classLastWrittenValueIndex, 1,
+            'Ignoring `class` binding because there is no `class` metadata associated with the element. ' +
+                '(Was exception thrown during `firstUpdatePass` which prevented the metadata creation?)');
+    ngDevMode &&
+        assertLessThan(classLastWrittenValueIndex, lView.length, 'Reading past end of LView');
```

Comment:
Nit: seems like we could use the "data in range" function we already have?

```suggestion
        assertDataInRange(lView, classLastWrittenValueIndex);
```